Code/Framework/AtomCore/AtomCore/Instance/InstanceId.h

```diff
@@ -65,7 +73,11 @@ namespace AZ
             void ToString(StringType& result) const;
 
             Uuid m_guid = Uuid::CreateNull();
-            uint32_t  m_subId = 0;
+            uint32_t m_subId = 0;
+
+            // Pointer used to uniquely identify a version of the data, usually an asset. It is not intended to be
+            // dereferenced or used to access asset data.
+            void* m_versionId = nullptr; 
```

Comment:
I still like tracking the reasoning behind the change with a comment like this. (I'm not blocking the review over this though).
```
// If AssetData is included as part of the ID, that will prevent the following situation: an asset is hot-reloaded
// and the OnAssetReloaded callback finds an old instance that uses the old asset instead of creating a
// new instance with the reloaded asset.
```

Reply:
In this case, if something like the material component receives the asset reloaded notification, it will create a new instance using the new asset data pointer as part of the ID, abandoning the old version of the instance, and send the notification to update material instance references. Another improvement that can be made after this is to give the instance database explicit functions for how to retrieve instances instead of creating the instance ID and passing it in. Then the instance database can be responsible for constructing the instance ID and returning it with the instance. This would allow the instance database to update the ID hash with the asset data pointer after it attempts to load the asset.